File: engine/src/main/java/com/terralite/engine/physics/Bounds.java
package com.terralite.engine.physics;

public record Bounds(double minX, double minY, double minZ, double maxX, double maxY, double maxZ) {
    public Bounds {
        if (maxX < minX) {
            throw new IllegalArgumentException("Max x cannot be less than min x");
        }
        if (maxY < minY) {
            throw new IllegalArgumentException("Max y cannot be less than min y");
        }
        if (maxZ < minZ) {
            throw new IllegalArgumentException("Max z cannot be less than min z");
        }
    }

    public Transform clamp(Transform transform) {
        return new Transform(
            clamp(transform.x(), minX, maxX),
            clamp(transform.y(), minY, maxY),
            clamp(transform.z(), minZ, maxZ)
        );
    }

    public boolean contains(Transform transform) {
        return transform.x() >= minX && transform.x() <= maxX
            && transform.y() >= minY && transform.y() <= maxY
            && transform.z() >= minZ && transform.z() <= maxZ;
    }

    private static double clamp(double value, double min, double max) {
        return Math.max(min, Math.min(max, value));
    }
}

File: engine/src/commonMain/kotlin/com/terralite/engine/physics/Bounds.kt
package com.terralite.engine.physics

@JvmRecord
data class Bounds(
    val minX: Double, val minY: Double, val minZ: Double,
    val maxX: Double, val maxY: Double, val maxZ: Double
) {
    init {
        require(maxX >= minX) { "Max x cannot be less than min x" }
        require(maxY >= minY) { "Max y cannot be less than min y" }
        require(maxZ >= minZ) { "Max z cannot be less than min z" }
    }

    fun clamp(transform: Transform): Transform = Transform(
        transform.x.coerceIn(minX, maxX),
        transform.y.coerceIn(minY, maxY),
        transform.z.coerceIn(minZ, maxZ)
    )

    fun contains(transform: Transform): Boolean =
        transform.x in minX..maxX &&
        transform.y in minY..maxY &&
        transform.z in minZ..maxZ
}
